diff --git a/src/platform/mpfs_hal/startup_gcc/mss_entry.S b/src/platform/mpfs_hal/startup_gcc/mss_entry.S
index 9e049de..e07ae56 100644
--- a/src/platform/mpfs_hal/startup_gcc/mss_entry.S
+++ b/src/platform/mpfs_hal/startup_gcc/mss_entry.S
@@ -55,7 +55,7 @@ _start:
     csrw medeleg, 0
 .skip_e51:
     # mscratch must be init to zero- we are not using scratch memory
-    csrw mscratch, zero
+   #csrw mscratch, zero
     csrw mcause, zero
     csrw mepc, zero
     /*
diff --git a/src/platform/mpfs_hal/startup_gcc/system_startup.c b/src/platform/mpfs_hal/startup_gcc/system_startup.c
index accd9ea..6bad203 100644
--- a/src/platform/mpfs_hal/startup_gcc/system_startup.c
+++ b/src/platform/mpfs_hal/startup_gcc/system_startup.c
@@ -456,7 +456,7 @@ __attribute__((weak)) uint8_t init_mem_protection_unit(void)
  */
 __attribute__((weak)) uint8_t init_pmp(uint8_t hart_id)
 {
-    pmp_configure(hart_id);
+  //pmp_configure(hart_id);
     return (0U);
 }
 